cantools/database/can/cpp_source.py

```diff
@@ -0,0 +1,710 @@
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+/**
+ * This file was generated by cantools version {version} {date}.
+ */
+
+#ifndef {include_guard}
+#define {include_guard}
+
+#include <string>
+#include <sstream>
+#include <iomanip>
+
+#include "absl/types/span.h"
+
+#include "DBC.h"
+
+{declarations}
+
+#endif  // {include_guard}
+'''
+
+SOURCE_FMT = '''\
+/**
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2018-2019 Erik Moqvist
+ *
+ * Permission is hereby granted, free of charge, to any person
+ * obtaining a copy of this software and associated documentation
+ * files (the "Software"), to deal in the Software without
+ * restriction, including without limitation the rights to use, copy,
+ * modify, merge, publish, distribute, sublicense, and/or sell copies
+ * of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+/**
+ * This file was generated by cantools version {version} {date}.
+ */
+
+#include "{header}"
+
+{definitions}\
+'''
+
+SIGNAL_DECLARATION_FMT = '''\
+/**
+ * Signal {name}.
+ * Message {message_name}
+ *
+{comment}\
+ * Range: {range}
+ * Scale: {scale}
+ * Offset: {offset}
+{additional_comments}\
+ */
+class {message_name}_{name} : public Signal<{type_name}, double> {{
+public:
+    {message_name}_{name}(const uint8_t* buffer);
+
+    virtual {type_name} Raw() override;
+    virtual bool RawInRange(const {type_name}& value) const override;
+}};
+'''
+
+MESSAGE_DECLARATION_FMT = '''\
+{signal_constructors}
+
+/**
+ * Message {database_message_name}.
+ *
+{comment}\
+ */
+class {database_message_name} : public Frame {{
+public:
+    // Constructor, empty buffer
+    {database_message_name}();
+    // Constructor to maintain ownership of input buffer
+    {database_message_name}(std::unique_ptr<uint8_t[]>&& other, const size_t size);
+    // Constructor pass in buffer and buffer size, user maintains buffer ownership after object destruction
+    {database_message_name}(uint8_t* buffer, const size_t size);
+
+    ~{database_message_name}() = default;
+
+    {database_message_name}& GetMessage() {{ 
```

Comment:
I'm not sure I understand the point of this function?  It doesn't seem to be conversion casting or anything, so why have it?  I could understand if this was a templated container, but it isn't.

Reply:
From our Slack convo:

> cause then the CanSlot class could take that class as a template parameter, and you can have a GetMessage method that would return a reference to the held class, and then knowing what type of message it is you just call the signal accessors

Reply:
Sorry, I was suggesting that the CanSlot class (which would be templatized on this) would have the GetMessage method, not the message class.